Create a document with the database schema (#7)
bosagora/stoa: docs/Database_Schema.md

| enrollment_count  | INTEGER   |    | Y        |   0      | The number of enrollments in the block|

### _Create Script_

```sql
CREATE TABLE "blocks" (
    "height"    INTEGER NOT NULL,
    "prev_block"    TEXT NOT NULL,
    "validators"    TEXT NOT NULL,
    "merkle_root"    TEXT NOT NULL,
    "signature"    TEXT NOT NULL,
    "tx_count"    INTEGER NOT NULL,
    "enrollment_count"    INTEGER NOT NULL DEFAULT 0,
    PRIMARY KEY("height")
)
```
----

## 2. Table **enrollments**

### _Schema_

| Column            | Data Type | PK | Not NULL | Default  |Description|
|:----------------- |:--------- |:--:|:--------:| -------- | --------- |
| block_height      | INTEGER   | Y  | Y        |          | The height of the block|
| enrollment_index  | INTEGER   | Y  | Y        |          | The sequence of enrollment in the block.|
| utxo_key          | TEXT      |    | Y        |          ||
| random_seed       | TEXT      |    | Y        |          ||
| cycle_length      | INTEGER   |    | Y        |          ||
| enroll_sig        | TEXT      |    | Y        |          ||

### _Create Script_

```sql
CREATE TABLE "enrollments" (
    "block_height"    INTEGER NOT NULL,

>     "enrollment_index"    INTEGER NOT NULL,

Reviewer: What is `enrollment_index` ?

Reviewer: The index of this `Enrollment` in the block's `enrollments` array

	"preimage_hash"	TEXT NOT NULL,
	PRIMARY KEY("block_height","utxo_key")
)
```
----

## 8. Table **accounts** `TBD`

### _Schema_

| Column             | Data Type | PK | Not NULL | Default  |Description|
|:------------------ |:--------- |:--:|:--------:| -------- | --------- |
|  block_height      | INTEGER   | Y  | Y        |          | The height of the block|
|  address           | TEXT(56)  | Y  | Y        |          | The public key |
|  amount            | NUMERIC   |    | Y        |          | spendable_amount + unspendable_amount |
|  spendable_amount  | NUMERIC   |    | Y        |          | The sum of the active(melted) UTXO |
|  unspendable_amount| NUMERIC   |    | Y        |          | The sum of the frozen + melting UTXO |
|  is_validator      | TEXT(1)   |    | Y        |          | Whether its owner is a validator ('Y': validator, 'N' : non-validator)|

### _Create Script_

```sql
CREATE TABLE "accounts" (
	"block_height"	INTEGER NOT NULL,
	"address"	TEXT(56) NOT NULL,
	"amount"	NUMERIC NOT NULL,
	"spendable_amount"	NUMERIC NOT NULL,
	"unspendable_amount"	NUMERIC NOT NULL
	"is_validator"	TEXT(1) NOT NULL
	PRIMARY KEY("block_height","address")
)

> ```

Reviewer: Let's leave this out of the current sprint.

|  output_index     | INTEGER   | Y  | Y        |          | The sequence of output in the outputs.|
|  utxo_key         | TEXT      |    | Y        |          ||
|  amount           | NUMERIC   |    | Y        |          ||
|  address          | TEXT(56)  |    | Y        |          ||

### _Create Script_

```sql
CREATE TABLE "tx_outputs" (
    "block_height"    INTEGER NOT NULL,
    "tx_index"    INTEGER NOT NULL,
    "output_index"    INTEGER NOT NULL,
    "utxo_key"    TEXT NOT NULL,
    "amount"    NUMERIC NOT NULL,
    "address"    TEXT(56) NOT NULL,
    PRIMARY KEY("block_height","tx_index","output_index")
)
```
----

## 6. Table **preimages**

### _Schema_

| Column            | Data Type | PK | Not NULL | Default  |Description|
|:----------------- |:--------- |:--:|:--------:| -------- | --------- |
|  utxo_key         | TEXT      | Y  | Y        |          ||
|  preimage_distance| INTEGER   |    | Y        |          ||
|  preimage_hash    | TEXT      |    | Y        |          ||

### _Create Script_

```sql
CREATE TABLE "preimages" (
	"utxo_key"	TEXT NOT NULL,
	"preimage_distance"	INTEGER NOT NULL,
	"preimage_hash"	INTEGER NOT NULL,
	PRIMARY KEY("utxo_key")
)

> ```

Reviewer: Remove

----

## 7. Table **validators**

### _Schema_

| Column            | Data Type | PK | Not NULL | Default  |Description|
|:----------------- |:--------- |:--:|:--------:| -------- | --------- |

> |  block_height     | INTEGER   | Y  | Y        |          | The height of the block|
> |  utxo_key         | TEXT      | Y  | Y        |          ||
> |  enrolled_at      | INTEGER   |    | Y        |          ||

Reviewer: ```suggestion
|  enrolled_at     | INTEGER   | Y  | Y        |          | The height this validator enrolled at |
|  utxo_key         | TEXT      | Y  | Y        |          ||
```
